File: config_builder.py
# -*- coding: utf-8 -*-
"""
Парсит ss:// ссылку и собирает JSON-конфиг для V2Ray/Xray на ПК.
Формат: ss://base64(method:password)@host:port#tag
"""
import base64
import json


def build_from_ss_link(ss_link: str) -> str | None:
    """Из ss:// ссылки собрать JSON-конфиг для V2Ray (с SOCKS и HTTP inbound для ПК)."""
    link = ss_link.strip().removeprefix("ss://")
    hash_idx = link.find("#")
    main = link[:hash_idx] if hash_idx >= 0 else link
    at_idx = main.find("@")
    if at_idx <= 0:
        return None
    user_info = main[:at_idx]
    host_port = main[at_idx + 1 :]
    colon = host_port.rfind(":")
    if colon <= 0:
        return None
    host = host_port[:colon]
    port_str = host_port[colon + 1 :].split("/")[0].split("?")[0]
    try:
        port = int(port_str)
    except ValueError:
        return None

    try:
        # URL-safe base64: - и _ заменяют + и /
        pad = user_info.replace("-", "+").replace("_", "/")
        pad += "=" * (4 - len(pad) % 4) if len(pad) % 4 else ""
        decoded = base64.b64decode(pad).decode("utf-8", errors="replace")
    except Exception:
        return None

    parts = decoded.split(":", 2)
    if len(parts) != 2:
        return None
    method = parts[0].strip()
    password = parts[1].strip()

    return _build_config(host, port, method, password)


def _build_config(host: str, port: int, method: str, password: str) -> str:
    """Конфиг с SOCKS 1081 и HTTP 3128 (для системного прокси ПК)."""
    config = {
        "log": {"loglevel": "warning"},
        "inbounds": [
            {
                "listen": "127.0.0.1",
                "port": 1081,
                "protocol": "socks",
                "tag": "socks-in",
                "settings": {"auth": "noauth", "udp": False},
            },
            {
                "listen": "127.0.0.1",
                "port": 3128,
                "protocol": "http",
                "tag": "http-in",
                "settings": {},
            },
        ],
        "outbounds": [
            {
                "protocol": "shadowsocks",
                "settings": {
                    "servers": [
                        {
                            "address": host,
                            "port": port,
                            "method": method,
                            "password": password,
                        }
                    ]
                },
                "tag": "proxy",
            },
            {"protocol": "freedom", "settings": {}, "tag": "direct"},
            {"protocol": "blackhole", "settings": {}, "tag": "block"},
        ],
        "routing": {
            "domainStrategy": "IPOnDemand",
            "rules": [
                {"type": "field", "ip": ["geoip:private"], "outboundTag": "block"},
                {"type": "field", "network": "tcp,udp", "outboundTag": "proxy"},
            ],
        },
    }
    return json.dumps(config, ensure_ascii=False, indent=2)

File: android-app/app/src/main/java/ru/vpnconfig/paste/ConfigBuilder.kt
package ru.vpnconfig.paste

import android.util.Base64

/**
 * Парсит ss:// ссылку и собирает JSON-конфиг для Xray/V2Ray.
 * Формат: ss://base64(method:password)@host:port#tag
 */
object ConfigBuilder {

    fun buildFromSsLink(ssLink: String): String? {
        val link = ssLink.trim().removePrefix("ss://")
        val hashIndex = link.indexOf('#')
        val main = if (hashIndex >= 0) link.substring(0, hashIndex) else link
        val atIndex = main.indexOf('@')
        if (atIndex <= 0) return null
        val userInfo = main.substring(0, atIndex)
        val hostPort = main.substring(atIndex + 1)
        val colon = hostPort.lastIndexOf(':')
        if (colon <= 0) return null
        val host = hostPort.substring(0, colon)
        val portStr = hostPort.substring(colon + 1).split("/").first().split("?").first()
        val port = portStr.toIntOrNull() ?: return null

        val decoded = try {
            String(Base64.decode(userInfo.replace("-", "+").replace("_", "/"), Base64.NO_WRAP or Base64.URL_SAFE))
        } catch (_: Exception) {
            return null
        }
        val methodPassword = decoded.split(":", limit = 2)
        if (methodPassword.size != 2) return null
        val method = methodPassword[0].trim()
        val password = methodPassword[1].trim()

        return buildConfig(host, port, method, password)
    }

    private fun buildConfig(host: String, port: Int, method: String, password: String): String {
        return """
{
  "log": { "loglevel": "warning" },
  "inbounds": [],
  "outbounds": [{
    "protocol": "shadowsocks",
    "settings": {
      "servers": [{
        "address": "$host",
        "port": $port,
        "method": "$method",
        "password": "$password"
      }]
    },
    "tag": "proxy"
  },{
    "protocol": "freedom",
    "settings": {},
    "tag": "direct"
  },{
    "protocol": "blackhole",
    "settings": {},
    "tag": "block"
  }],
  "routing": {
    "domainStrategy": "IPOnDemand",
    "rules": [
      { "type": "field", "network": "tcp,udp", "outboundTag": "proxy" }
    ]
  }
}
        """.trimIndent()
    }
}
